### ISSUE TITLE
IndexOutOfRangeException, EOMapRenderer.cs, Line 344, MapRef.WarpLookup[destY, destX];

### ISSUE THREAD

**Issue by squiblez** [ethanmoffat/EndlessClient#37]
This just started as of late, I decided to retry adding my music engine and bam! I can login onto any map just fine(most of the time), but no matter what server I use(including my own test server, using the normal EO maps)- It always throws this exception after I try moving to another map.

I figured it would be best to let you know instead of just ignore it and hope for the best haha.

It might also help to know that, this was not an issue I experienced when compiling a version of the code from about a week ago.

![mapref warplookup 9 10 2015](https://cloud.githubusercontent.com/assets/14168449/9782480/686bf8b6-576b-11e5-91d0-120c96e924ed.png)
![screenshot23](https://cloud.githubusercontent.com/assets/14168449/9782503/87bdf5de-576b-11e5-9b92-8eaeb29aa4f5.jpg)

**Comment by ethanmoffat** [comment]
This is probably related to the recent commit where I split input handling out of EOCharacterRenderer and into separate game components. I've only encountered this once since that change and thought it was a fluke. I'll have a fix in for it soon.

**Comment by squiblez** [comment]
The IndexOutOfRange exception can happen anywhere, especially on large maps! It's a time related bug. Forcing the code to wait or lag if DestX and DestY are larger than the appropriate dimension BEFORE the exception would be thrown prevents it- however this causes a lot of render issues(character disappearing and jumping back and forth), and if this happens multiple times it will eventually crash the client with no exceptions.

I'm guessing this all has to do with loading and/or downloading the map while it's trying to draw. Possibly halting the renderer for a while could solve fix it. If you are looking for a fast way to throw the exception for testing, walk into the Aeven grocer, and immediately walk out, and then walk back in. Doing this 3-4 times is a sure-fire way.

Since your last update, throwing the except seldom breaks the game unless it happens back-to-back.

**Comment by ethanmoffat** [comment]
I've got a better / more permanent fix that I haven't submitted yet. 

The problem occurs as a result of trying to validate the 'next' tile on a walk action when the map is changing. The next tile is out of bounds of the map size - this used to be automatically checked by the way input was handled and broke since I changed it to be more modular.

I tried adding a synchronization object to prevent trying to validate the next tile while the map is updating but I think that adds too much unnecessary overhead (and it wasn't working properly anyway). I've added bounds checking for GetTileInfo that should take care of this problem in all cases.

I am currently optimizing the fix (I'm catching the IndexOutOfRangeExceptions currently and doing nothing on that error, which is a band-aid at best) and testing that it still works with the bounds checking method and then this issue should be taken care of.
